src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol_ImplementationChecks.cs

```diff
@@ -1025,10 +1016,94 @@ private void CheckNewModifier(Symbol symbol, bool isNew, DiagnosticBag diagnosti
             }
 
+            return;
+
             bool isOrContainsErrorType(TypeSymbol typeSymbol)
             {
                 return (object)typeSymbol.VisitType((currentTypeSymbol, unused1, unused2) => currentTypeSymbol.IsErrorType(), (object)null) != null;
             }
+
+            static void checkValidNullableMethodOverride(
+                Location overridingMemberLocation,
+                MethodSymbol overriddenMethod,
+                MethodSymbol overridingMethod,
+                DiagnosticBag diagnostics,
+                bool checkParameters)
+            {
+                var compilation = overridingMethod.DeclaringCompilation;
+                if (overriddenMethod is null ||
+                    overridingMethod is null ||
+                    compilation is null ||
+                    !compilation.IsFeatureEnabled(MessageID.IDS_FeatureNullableReferenceTypes))
+                {
+                    // Don't do any validation if the nullable feature is not enabled or
+                    // the override is not written directly in source
+                    return;
+                }
+
+                var conversions = compilation.Conversions.WithNullability(true);
+                if (!isValidNullableConversion(
+                        conversions,
+                        overridingMethod.RefKind,
+                        overridingMethod.ReturnTypeWithAnnotations,
+                        overriddenMethod.ReturnTypeWithAnnotations))
+                {
+                    diagnostics.Add(ErrorCode.WRN_NullabilityMismatchInReturnTypeOnOverride, overridingMemberLocation);
+                    return;
+                }
+
+                if (!checkParameters)
+                {
+                    return;
+                }
+
+                ImmutableArray<ParameterSymbol> overridingParameters = overridingMethod.GetParameters();
+                var overriddenParameters = overriddenMethod.GetParameters();
+
+                for (int i = 0; i < overriddenMethod.ParameterCount; i++)
+                {
+                    var overriddenParameterType = overriddenParameters[i].TypeWithAnnotations;
+                    var overridingParameterType = overridingParameters[i].TypeWithAnnotations;
+                    if (!isValidNullableConversion(
+                            conversions,
+                            overridingParameters[i].RefKind,
+                            overriddenParameterType,
+                            overridingParameterType))
+                    {
+                        diagnostics.Add(
+                            ErrorCode.WRN_NullabilityMismatchInParameterTypeOnOverride,
+                            overridingMemberLocation,
+                            new FormattedSymbol(overridingParameters[i], SymbolDisplayFormat.ShortFormat));
+                    }
+                }
+
+                return;
+
+                static bool isValidNullableConversion(
```

Comment:
>isValidNullableConversion [](start = 28, length = 25)

If you don't mind, could we move this local function out one level, to avoid nested local functions?